File: dotnet/tests/Obscura.Cdp.Tests/InputMouseEventParity.cs
using System.Globalization;
using System.Text.Json.Nodes;
using Obscura.Cdp;
using Xunit;

namespace Obscura.Cdp.Tests;

/// <summary>
/// The xUnit port of <c>crates/obscura-cdp/tests/input_mouse_event_parity.rs</c>.
/// </summary>
[Collection(CdpDomainCollection.Name)]
public sealed class InputMouseEventParity
{
    private static CdpTestServer ServeFixture() => CdpTestServer.ServeHtml(
        """
        <!doctype html><html><head><style>
                    html, body { margin: 0; }
                    #page { width: 1800px; height: 2400px; }
                    #box { position: absolute; left: 20px; top: 20px; width: 180px;
                           height: 120px; overflow: auto; border: 10px solid black; }
                    #inner { width: 700px; height: 800px; }
                </style></head><body>
                  <div id="page"></div>
                  <div id="box"><div id="inner"></div></div>
                  <input id="check" type="checkbox">
                  <form id="radio-form">
                    <input id="radio-a" type="radio" name="choice" checked>
                    <input id="radio-b" type="radio" name="choice">
                  </form>
                </body></html>
        """);

    private static async Task<JsonNode> CdpAsync(
        CdpContext ctx,
        ulong id,
        string method,
        string parameters,
        string sessionId)
    {
        CdpResponse response = await Dispatcher.DispatchAsync(
            new CdpRequest
            {
                Id = id,
                Method = method,
                Params = CdpDomainFixtures.Json(parameters),
                SessionId = sessionId,
            },
            ctx);
        Assert.True(response.Error is null, $"CDP {method} failed: {response.Error?.Message}");
        return response.Result ?? new JsonObject();
    }

    private static Task<JsonNode> EvaluateAsync(
        CdpContext ctx,
        ulong id,
        string expression,
        string sessionId) =>
        CdpAsync(
            ctx,
            id,
            "Runtime.evaluate",
            $$"""
            {
                "expression": {{CdpJson.String(expression)}},
                "returnByValue": true,
                "awaitPromise": true
            }
            """,
            sessionId);

    private static async Task<(CdpContext Ctx, string SessionId, CdpTestServer Server)> SetupAsync()
    {
        CdpTestServer server = ServeFixture();
        var ctx = CdpContext.New();
        string pageId = ctx.CreatePage();
        const string SessionId = "input-mouse-session";
        ctx.Sessions[SessionId] = pageId;
        await CdpAsync(
            ctx,
            1,
            "Page.navigate",
            $$"""{"url": "{{server.Url}}", "waitUntil": "load"}""",
            SessionId);
        return (ctx, SessionId, server);
    }

    private static Task WheelAsync(
        CdpContext ctx,
        ulong id,
        string sessionId,
        double x,
        double y,
        double deltaX,
        double deltaY) =>
        CdpAsync(
            ctx,
            id,
            "Input.dispatchMouseEvent",
            string.Create(
                CultureInfo.InvariantCulture,
                $$"""
                {"type": "mouseWheel", "x": {{x}}, "y": {{y}},
                 "deltaX": {{deltaX}}, "deltaY": {{deltaY}}}
                """),
            sessionId);

    private static async Task<JsonNode> ScrollStateAsync(CdpContext ctx, ulong id, string sessionId)
    {
        JsonNode result = await EvaluateAsync(
            ctx,
            id,
            """
            JSON.stringify({
                rootX: scrollX, rootY: scrollY,
                boxX: document.getElementById('box').scrollLeft,
                boxY: document.getElementById('box').scrollTop,
                rootScrollWidth: document.scrollingElement.scrollWidth,
                rootClientWidth: document.scrollingElement.clientWidth,
                pageRect: document.getElementById('page').getBoundingClientRect().toJSON(),
                maxBoxX: document.getElementById('box').scrollWidth - document.getElementById('box').clientWidth,
                maxBoxY: document.getElementById('box').scrollHeight - document.getElementById('box').clientHeight
            })
            """,
            sessionId);
        return CdpDomainFixtures.Json(result["result"]!["value"]!.GetValue<string>());
    }

    private static double Number(JsonNode? node) => node!.GetValue<double>();

    [Fact]
    public async Task WheelOverPageScrollsTheRootOnBothAxes()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            await WheelAsync(ctx, 2, sid, 600.0, 300.0, 45.0, 160.0);
            JsonNode state = await ScrollStateAsync(ctx, 3, sid);
            Assert.Equal(45.0, Number(state["rootX"]));
            Assert.Equal(160.0, Number(state["rootY"]));
            Assert.Equal(0.0, Number(state["boxX"]));
            Assert.Equal(0.0, Number(state["boxY"]));
        }
    }

    [Fact]
    public async Task WheelOverNestedOverflowScrollsTheNestedContainer()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            await WheelAsync(ctx, 2, sid, 50.0, 50.0, 70.0, 110.0);
            JsonNode state = await ScrollStateAsync(ctx, 3, sid);
            Assert.Equal(70.0, Number(state["boxX"]));
            Assert.Equal(110.0, Number(state["boxY"]));
            Assert.Equal(0.0, Number(state["rootX"]));
            Assert.Equal(0.0, Number(state["rootY"]));
        }
    }

    [Fact]
    public async Task WheelOffsetsClampToNestedScrollExtents()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            await WheelAsync(ctx, 2, sid, 50.0, 50.0, 100000.0, 100000.0);
            JsonNode state = await ScrollStateAsync(ctx, 3, sid);
            Assert.Equal(Number(state["maxBoxX"]), Number(state["boxX"]));
            Assert.Equal(Number(state["maxBoxY"]), Number(state["boxY"]));

            await WheelAsync(ctx, 4, sid, 50.0, 50.0, -100000.0, -100000.0);
            state = await ScrollStateAsync(ctx, 5, sid);
            Assert.Equal(0.0, Number(state["boxX"]));
            Assert.Equal(0.0, Number(state["boxY"]));
        }
    }

    [Fact]
    public async Task WheelChainsToRootWhenNestedScrollerIsSaturated()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            await EvaluateAsync(
                ctx,
                2,
                "(() => { const box = document.getElementById('box'); box.scrollTop = box.scrollHeight; })()",
                sid);
            JsonNode saturated = await ScrollStateAsync(ctx, 3, sid);
            Assert.Equal(Number(saturated["maxBoxY"]), Number(saturated["boxY"]));

            await WheelAsync(ctx, 4, sid, 50.0, 50.0, 0.0, 90.0);
            JsonNode state = await ScrollStateAsync(ctx, 5, sid);
            Assert.Equal(Number(state["maxBoxY"]), Number(state["boxY"]));
            Assert.Equal(90.0, Number(state["rootY"]));
        }
    }

    [Fact]
    public async Task CancelingWheelPreventsItsScrollDefault()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            await EvaluateAsync(
                ctx,
                2,
                """
                (() => {
                    globalThis.wheelProbe = null;
                    const page = document.getElementById('page');
                    document.elementFromPoint = () => page;
                    page.addEventListener('wheel', event => {
                        wheelProbe = {
                            x: event.clientX, y: event.clientY,
                            dx: event.deltaX, dy: event.deltaY,
                            ctrl: event.ctrlKey, trusted: event.isTrusted
                        };
                        event.preventDefault();
                    });
                })()
                """,
                sid);
            await CdpAsync(
                ctx,
                3,
                "Input.dispatchMouseEvent",
                """
                {
                    "type": "mouseWheel", "x": 600.0, "y": 300.0,
                    "deltaX": 25.0, "deltaY": 75.0, "modifiers": 2
                }
                """,
                sid);
            JsonNode state = await ScrollStateAsync(ctx, 4, sid);
            Assert.Equal(0.0, Number(state["rootX"]));
            Assert.Equal(0.0, Number(state["rootY"]));

            JsonNode raw = await EvaluateAsync(ctx, 5, "JSON.stringify(wheelProbe)", sid);
            JsonNode probe = CdpDomainFixtures.Json(raw["result"]!["value"]!.GetValue<string>());
            Assert.Equal(600.0, Number(probe["x"]));
            Assert.Equal(300.0, Number(probe["y"]));
            Assert.Equal(25.0, Number(probe["dx"]));
            Assert.Equal(75.0, Number(probe["dy"]));
            Assert.True(probe["ctrl"]!.GetValue<bool>());
            Assert.True(probe["trusted"]!.GetValue<bool>());
        }
    }

    [Fact]
    public async Task HitTestingClipsScrolledChildrenAtOverflowPaddingEdge()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            JsonNode raw = await EvaluateAsync(
                ctx,
                2,
                """
                (() => {
                    const box = document.getElementById('box');
                    box.scrollLeft = 50;
                    const inner = document.getElementById('inner').getBoundingClientRect();
                    return JSON.stringify({
                        hit: document.elementFromPoint(25, 50).id,
                        innerLeft: inner.left, innerRight: inner.right,
                        boxLeft: box.getBoundingClientRect().left
                    });
                })()
                """,
                sid);
            JsonNode result = CdpDomainFixtures.Json(raw["result"]!["value"]!.GetValue<string>());
            Assert.True(Number(result["innerLeft"]) <= 25.0);
            Assert.True(Number(result["innerRight"]) >= 25.0);
            Assert.Equal(20.0, Number(result["boxLeft"]));
            Assert.Equal("box", result["hit"]!.GetValue<string>());
        }
    }

    [Fact]
    public async Task PressReleaseOrdersEventsAndDefersClickActivation()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            await EvaluateAsync(
                ctx,
                2,
                """
                (() => {
                    const target = document.getElementById('check');
                    document.elementFromPoint = () => target;
                    globalThis.mouseLog = [];
                    for (const type of ['mousedown', 'mouseup', 'click', 'input', 'change']) {
                        target.addEventListener(type, event => mouseLog.push({
                            type, checked: target.checked, x: event.clientX,
                            ctrl: event.ctrlKey, shift: event.shiftKey, trusted: event.isTrusted
                        }));
                    }
                })()
                """,
                sid);

            await CdpAsync(
                ctx,
                3,
                "Input.dispatchMouseEvent",
                """
                {
                    "type": "mousePressed", "x": 31.0, "y": 42.0,
                    "button": "left", "clickCount": 1, "modifiers": 10
                }
                """,
                sid);
            JsonNode pressedRaw = await EvaluateAsync(
                ctx,
                4,
                "JSON.stringify({log: mouseLog, checked: document.getElementById('check').checked})",
                sid);
            JsonNode pressed = CdpDomainFixtures.Json(
                pressedRaw["result"]!["value"]!.GetValue<string>());
            Assert.False(pressed["checked"]!.GetValue<bool>());
            Assert.Equal("mousedown", pressed["log"]![0]!["type"]!.GetValue<string>());
            Assert.Single(pressed["log"]!.AsArray());

            await CdpAsync(
                ctx,
                5,
                "Input.dispatchMouseEvent",
                """
                {
                    "type": "mouseReleased", "x": 31.0, "y": 42.0,
                    "button": "left", "clickCount": 1, "modifiers": 10
                }
                """,
                sid);
            JsonNode releasedRaw = await EvaluateAsync(
                ctx,
                6,
                "JSON.stringify({log: mouseLog, checked: document.getElementById('check').checked})",
                sid);
            JsonNode released = CdpDomainFixtures.Json(
                releasedRaw["result"]!["value"]!.GetValue<string>());
            List<string> types =
                [.. released["log"]!.AsArray().Select(entry => entry!["type"]!.GetValue<string>())];
            Assert.Equal(["mousedown", "mouseup", "click", "input", "change"], types);
            Assert.True(released["checked"]!.GetValue<bool>());
            Assert.True(released["log"]![2]!["checked"]!.GetValue<bool>());
            Assert.Equal(31.0, Number(released["log"]![2]!["x"]));
            Assert.True(released["log"]![2]!["ctrl"]!.GetValue<bool>());
            Assert.True(released["log"]![2]!["shift"]!.GetValue<bool>());
            Assert.True(released["log"]![2]!["trusted"]!.GetValue<bool>());
        }
    }

    [Fact]
    public async Task RadioReleaseSelectsOnlyTheTargetInItsGroup()
    {
        (CdpContext ctx, string sid, CdpTestServer server) = await SetupAsync();
        using (server)
        {
            await EvaluateAsync(
                ctx,
                2,
                """
                (() => {
                    const a = document.getElementById('radio-a');
                    const b = document.getElementById('radio-b');
                    document.elementFromPoint = () => b;
                    globalThis.radioEvents = [];
                    for (const radio of [a, b]) {
                        for (const type of ['mousedown', 'mouseup', 'click', 'input', 'change']) {
                            radio.addEventListener(type, () => radioEvents.push(radio.id + ':' + type));
                        }
                    }
                })()
                """,
                sid);
            await CdpAsync(
                ctx,
                3,
                "Input.dispatchMouseEvent",
                """{"type": "mousePressed", "x": 10.0, "y": 10.0, "button": "left"}""",
                sid);
            await CdpAsync(
                ctx,
                4,
                "Input.dispatchMouseEvent",
                """{"type": "mouseReleased", "x": 10.0, "y": 10.0, "button": "left"}""",
                sid);
            JsonNode raw = await EvaluateAsync(
                ctx,
                5,
                "JSON.stringify({a: document.getElementById('radio-a').checked,"
                + " b: document.getElementById('radio-b').checked, events: radioEvents})",
                sid);
            JsonNode result = CdpDomainFixtures.Json(raw["result"]!["value"]!.GetValue<string>());
            Assert.False(result["a"]!.GetValue<bool>());
            Assert.True(result["b"]!.GetValue<bool>());
            CdpDomainFixtures.AssertJson(
                """
                ["radio-b:mousedown", "radio-b:mouseup", "radio-b:click", "radio-b:input",
                 "radio-b:change"]
                """,
                result["events"]);
        }
    }
}

File: .reference/obscura/crates/obscura-cdp/tests/input_mouse_event_parity.rs
#![cfg(feature = "render")]

use obscura_cdp::dispatch::{dispatch, CdpContext};
use obscura_cdp::types::CdpRequest;
use serde_json::{json, Value};
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpListener;

async fn serve_fixture() -> String {
    let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
    let addr = listener.local_addr().unwrap();
    tokio::spawn(async move {
        let (mut socket, _) = listener.accept().await.unwrap();
        let mut buf = [0u8; 2048];
        let _ = socket.read(&mut buf).await.unwrap();
        let body = r#"<!doctype html><html><head><style>
            html, body { margin: 0; }
            #page { width: 1800px; height: 2400px; }
            #box { position: absolute; left: 20px; top: 20px; width: 180px;
                   height: 120px; overflow: auto; border: 10px solid black; }
            #inner { width: 700px; height: 800px; }
        </style></head><body>
          <div id="page"></div>
          <div id="box"><div id="inner"></div></div>
          <input id="check" type="checkbox">
          <form id="radio-form">
            <input id="radio-a" type="radio" name="choice" checked>
            <input id="radio-b" type="radio" name="choice">
          </form>
        </body></html>"#;
        let response = format!(
            "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{body}",
            body.len()
        );
        let _ = socket.write_all(response.as_bytes()).await;
    });
    format!("http://{addr}/")
}

async fn cdp(
    ctx: &mut CdpContext,
    id: u64,
    method: &str,
    params: Value,
    session_id: &str,
) -> Value {
    let response = dispatch(
        &CdpRequest {
            id,
            method: method.to_string(),
            params,
            session_id: Some(session_id.to_string()),
        },
        ctx,
    )
    .await;
    assert!(response.error.is_none(), "CDP {method} failed: {:?}", response.error);
    response.result.unwrap_or_else(|| json!({}))
}

async fn evaluate(ctx: &mut CdpContext, id: u64, expression: &str, session_id: &str) -> Value {
    cdp(
        ctx,
        id,
        "Runtime.evaluate",
        json!({"expression": expression, "returnByValue": true, "awaitPromise": true}),
        session_id,
    )
    .await
}

async fn setup() -> (CdpContext, String) {
    std::env::set_var("OBSCURA_ALLOW_PRIVATE_NETWORK", "1");
    let url = serve_fixture().await;
    let mut ctx = CdpContext::new();
    let page_id = ctx.create_page();
    let session_id = "input-mouse-session";
    ctx.sessions.insert(session_id.to_string(), page_id);
    cdp(
        &mut ctx,
        1,
        "Page.navigate",
        json!({"url": url, "waitUntil": "load"}),
        session_id,
    )
    .await;
    (ctx, session_id.to_string())
}

async fn wheel(ctx: &mut CdpContext, id: u64, sid: &str, x: f64, y: f64, dx: f64, dy: f64) {
    cdp(
        ctx,
        id,
        "Input.dispatchMouseEvent",
        json!({"type": "mouseWheel", "x": x, "y": y, "deltaX": dx, "deltaY": dy}),
        sid,
    )
    .await;
}

async fn scroll_state(ctx: &mut CdpContext, id: u64, sid: &str) -> Value {
    let result = evaluate(
        ctx,
        id,
        r#"JSON.stringify({
            rootX: scrollX, rootY: scrollY,
            boxX: document.getElementById('box').scrollLeft,
            boxY: document.getElementById('box').scrollTop,
            rootScrollWidth: document.scrollingElement.scrollWidth,
            rootClientWidth: document.scrollingElement.clientWidth,
            pageRect: document.getElementById('page').getBoundingClientRect().toJSON(),
            maxBoxX: document.getElementById('box').scrollWidth - document.getElementById('box').clientWidth,
            maxBoxY: document.getElementById('box').scrollHeight - document.getElementById('box').clientHeight
        })"#,
        sid,
    )
    .await;
    serde_json::from_str(result["result"]["value"].as_str().unwrap()).unwrap()
}

#[tokio::test(flavor = "current_thread")]
async fn wheel_over_page_scrolls_the_root_on_both_axes() {
    let (mut ctx, sid) = setup().await;
    wheel(&mut ctx, 2, &sid, 600.0, 300.0, 45.0, 160.0).await;
    let state = scroll_state(&mut ctx, 3, &sid).await;
    assert_eq!(state["rootX"], 45.0, "unexpected root geometry: {state}");
    assert_eq!(state["rootY"], 160.0);
    assert_eq!(state["boxX"], 0.0);
    assert_eq!(state["boxY"], 0.0);
}

#[tokio::test(flavor = "current_thread")]
async fn wheel_over_nested_overflow_scrolls_the_nested_container() {
    let (mut ctx, sid) = setup().await;
    wheel(&mut ctx, 2, &sid, 50.0, 50.0, 70.0, 110.0).await;
    let state = scroll_state(&mut ctx, 3, &sid).await;
    assert_eq!(state["boxX"], 70.0);
    assert_eq!(state["boxY"], 110.0);
    assert_eq!(state["rootX"], 0.0, "nested wheel must not leak to the viewport");
    assert_eq!(state["rootY"], 0.0, "nested wheel must not leak to the viewport");
}

#[tokio::test(flavor = "current_thread")]
async fn wheel_offsets_clamp_to_nested_scroll_extents() {
    let (mut ctx, sid) = setup().await;
    wheel(&mut ctx, 2, &sid, 50.0, 50.0, 100_000.0, 100_000.0).await;
    let state = scroll_state(&mut ctx, 3, &sid).await;
    assert_eq!(state["boxX"], state["maxBoxX"]);
    assert_eq!(state["boxY"], state["maxBoxY"]);

    wheel(&mut ctx, 4, &sid, 50.0, 50.0, -100_000.0, -100_000.0).await;
    let state = scroll_state(&mut ctx, 5, &sid).await;
    assert_eq!(state["boxX"], 0.0);
    assert_eq!(state["boxY"], 0.0);
}

#[tokio::test(flavor = "current_thread")]
async fn wheel_chains_to_root_when_nested_scroller_is_saturated() {
    let (mut ctx, sid) = setup().await;
    evaluate(
        &mut ctx,
        2,
        "(() => { const box = document.getElementById('box'); box.scrollTop = box.scrollHeight; })()",
        &sid,
    )
    .await;
    let saturated = scroll_state(&mut ctx, 3, &sid).await;
    assert_eq!(saturated["boxY"], saturated["maxBoxY"]);

    wheel(&mut ctx, 4, &sid, 50.0, 50.0, 0.0, 90.0).await;
    let state = scroll_state(&mut ctx, 5, &sid).await;
    assert_eq!(state["boxY"], state["maxBoxY"], "inner remains clamped");
    assert_eq!(state["rootY"], 90.0, "remaining wheel gesture chains to the viewport");
}

#[tokio::test(flavor = "current_thread")]
async fn canceling_wheel_prevents_its_scroll_default() {
    let (mut ctx, sid) = setup().await;
    evaluate(
        &mut ctx,
        2,
        r#"(() => {
            globalThis.wheelProbe = null;
            const page = document.getElementById('page');
            document.elementFromPoint = () => page;
            page.addEventListener('wheel', event => {
                wheelProbe = {
                    x: event.clientX, y: event.clientY,
                    dx: event.deltaX, dy: event.deltaY,
                    ctrl: event.ctrlKey, trusted: event.isTrusted
                };
                event.preventDefault();
            });
        })()"#,
        &sid,
    )
    .await;
    cdp(
        &mut ctx,
        3,
        "Input.dispatchMouseEvent",
        json!({
            "type": "mouseWheel", "x": 600.0, "y": 300.0,
            "deltaX": 25.0, "deltaY": 75.0, "modifiers": 2
        }),
        &sid,
    )
    .await;
    let state = scroll_state(&mut ctx, 4, &sid).await;
    assert_eq!(state["rootX"], 0.0);
    assert_eq!(state["rootY"], 0.0);
    let probe = evaluate(&mut ctx, 5, "JSON.stringify(wheelProbe)", &sid).await;
    let probe: Value = serde_json::from_str(probe["result"]["value"].as_str().unwrap()).unwrap();
    assert_eq!(probe["x"], 600.0);
    assert_eq!(probe["y"], 300.0);
    assert_eq!(probe["dx"], 25.0);
    assert_eq!(probe["dy"], 75.0);
    assert_eq!(probe["ctrl"], true);
    assert_eq!(probe["trusted"], true);
}

#[tokio::test(flavor = "current_thread")]
async fn hit_testing_clips_scrolled_children_at_overflow_padding_edge() {
    let (mut ctx, sid) = setup().await;
    let result = evaluate(
        &mut ctx,
        2,
        r#"(() => {
            const box = document.getElementById('box');
            box.scrollLeft = 50;
            const inner = document.getElementById('inner').getBoundingClientRect();
            return JSON.stringify({
                hit: document.elementFromPoint(25, 50).id,
                innerLeft: inner.left, innerRight: inner.right,
                boxLeft: box.getBoundingClientRect().left
            });
        })()"#,
        &sid,
    )
    .await;
    let result: Value = serde_json::from_str(result["result"]["value"].as_str().unwrap()).unwrap();
    assert!(result["innerLeft"].as_f64().unwrap() <= 25.0);
    assert!(result["innerRight"].as_f64().unwrap() >= 25.0);
    assert_eq!(result["boxLeft"], 20.0);
    assert_eq!(result["hit"], "box", "content hidden behind the border cannot win hit testing");
}

#[tokio::test(flavor = "current_thread")]
async fn press_release_orders_events_and_defers_click_activation() {
    let (mut ctx, sid) = setup().await;
    evaluate(
        &mut ctx,
        2,
        r#"(() => {
            const target = document.getElementById('check');
            document.elementFromPoint = () => target;
            globalThis.mouseLog = [];
            for (const type of ['mousedown', 'mouseup', 'click', 'input', 'change']) {
                target.addEventListener(type, event => mouseLog.push({
                    type, checked: target.checked, x: event.clientX,
                    ctrl: event.ctrlKey, shift: event.shiftKey, trusted: event.isTrusted
                }));
            }
        })()"#,
        &sid,
    )
    .await;

    cdp(
        &mut ctx,
        3,
        "Input.dispatchMouseEvent",
        json!({
            "type": "mousePressed", "x": 31.0, "y": 42.0,
            "button": "left", "clickCount": 1, "modifiers": 10
        }),
        &sid,
    )
    .await;
    let pressed = evaluate(
        &mut ctx,
        4,
        "JSON.stringify({log: mouseLog, checked: document.getElementById('check').checked})",
        &sid,
    )
    .await;
    let pressed: Value = serde_json::from_str(pressed["result"]["value"].as_str().unwrap()).unwrap();
    assert_eq!(pressed["checked"], false, "checkbox activation must wait for release");
    assert_eq!(pressed["log"][0]["type"], "mousedown");
    assert_eq!(pressed["log"].as_array().unwrap().len(), 1, "press must not synthesize click");

    cdp(
        &mut ctx,
        5,
        "Input.dispatchMouseEvent",
        json!({
            "type": "mouseReleased", "x": 31.0, "y": 42.0,
            "button": "left", "clickCount": 1, "modifiers": 10
        }),
        &sid,
    )
    .await;
    let released = evaluate(
        &mut ctx,
        6,
        "JSON.stringify({log: mouseLog, checked: document.getElementById('check').checked})",
        &sid,
    )
    .await;
    let released: Value = serde_json::from_str(released["result"]["value"].as_str().unwrap()).unwrap();
    let types: Vec<&str> = released["log"]
        .as_array()
        .unwrap()
        .iter()
        .map(|entry| entry["type"].as_str().unwrap())
        .collect();
    assert_eq!(types, ["mousedown", "mouseup", "click", "input", "change"]);
    assert_eq!(released["checked"], true);
    assert_eq!(released["log"][2]["checked"], true, "click sees checkbox pre-activation");
    assert_eq!(released["log"][2]["x"], 31.0);
    assert_eq!(released["log"][2]["ctrl"], true);
    assert_eq!(released["log"][2]["shift"], true);
    assert_eq!(released["log"][2]["trusted"], true);
}

#[tokio::test(flavor = "current_thread")]
async fn radio_release_selects_only_the_target_in_its_group() {
    let (mut ctx, sid) = setup().await;
    evaluate(
        &mut ctx,
        2,
        r#"(() => {
            const a = document.getElementById('radio-a');
            const b = document.getElementById('radio-b');
            document.elementFromPoint = () => b;
            globalThis.radioEvents = [];
            for (const radio of [a, b]) {
                for (const type of ['mousedown', 'mouseup', 'click', 'input', 'change']) {
                    radio.addEventListener(type, () => radioEvents.push(radio.id + ':' + type));
                }
            }
        })()"#,
        &sid,
    )
    .await;
    cdp(
        &mut ctx,
        3,
        "Input.dispatchMouseEvent",
        json!({"type": "mousePressed", "x": 10.0, "y": 10.0, "button": "left"}),
        &sid,
    )
    .await;
    cdp(
        &mut ctx,
        4,
        "Input.dispatchMouseEvent",
        json!({"type": "mouseReleased", "x": 10.0, "y": 10.0, "button": "left"}),
        &sid,
    )
    .await;
    let result = evaluate(
        &mut ctx,
        5,
        "JSON.stringify({a: document.getElementById('radio-a').checked, b: document.getElementById('radio-b').checked, events: radioEvents})",
        &sid,
    )
    .await;
    let result: Value = serde_json::from_str(result["result"]["value"].as_str().unwrap()).unwrap();
    assert_eq!(result["a"], false);
    assert_eq!(result["b"], true);
    assert_eq!(
        result["events"],
        json!(["radio-b:mousedown", "radio-b:mouseup", "radio-b:click", "radio-b:input", "radio-b:change"]),
        "the newly selected radio alone receives activation events"
    );
}
